sim/mcuboot-sys/src/api.rs

```diff
@@ -7,97 +7,202 @@ use log::{Level, log_enabled, warn};
+    align: u8,
+    erased_val: u8,
+}
+
+pub type FlashParams = HashMap<u8, FlashParamsStruct>;
+
+pub struct CAreaDescPtr {
+   pub ptr: *const CAreaDesc,
+}
+unsafe impl Send for CAreaDescPtr {}
+
+struct FlashContext {
+    flash_map: FlashMap,
+    flash_params: FlashParams,
+    flash_areas: CAreaDescPtr,
+}
+
+impl FlashContext {
+    pub fn new() -> FlashContext {
+        FlashContext {
+            flash_map: HashMap::new(),
+            flash_params: HashMap::new(),
+            flash_areas: CAreaDescPtr{ptr: ptr::null()},
+        }
+    }
+}
+
 lazy_static! {
-    static ref FLASH: Mutex<FlashMap> = {
+    static ref THREAD_CTX: Mutex<HashMap<thread::ThreadId, FlashContext>> = {
         Mutex::new(HashMap::new())
     };
 }
 
-struct FlashParams {
-    align: u8,
-    erased_val: u8,
+#[repr(C)]
+#[derive(Debug, Default)]
+pub struct CSimContext {
+    pub flash_counter: libc::c_int,
+    pub jumped: libc::c_int,
+    pub c_asserts: u8,
+    pub c_catch_asserts: u8,
+    // NOTE: Always leave boot_jmpbuf declaration at the end; this should
+    // store a "jmp_buf" which is arch specific and not defined by libc crate.
+    // The size below is enough to store data on a x86_64 machine.
+    pub boot_jmpbuf: [u64; 16],
+}
+
+pub struct CSimContextPtr {
+   pub ptr: *const CSimContext,
 }
+unsafe impl Send for CSimContextPtr {}
 
 lazy_static! {
-    static ref FLASH_PARAMS: Mutex<HashMap<u8, FlashParams>> = {
+    static ref SIM_CONTEXT: Mutex<HashMap<thread::ThreadId, CSimContextPtr>> = {
         Mutex::new(HashMap::new())
     };
 }
 
 // Set the flash device to be used by the simulation.  The pointer is unsafely stashed away.
 pub unsafe fn set_flash(dev_id: u8, dev: &mut dyn Flash) {
-    let mut flash_params = FLASH_PARAMS.lock().unwrap();
-    flash_params.insert(dev_id, FlashParams {
+    let mut map = THREAD_CTX.lock().unwrap();
+    let thread_id = thread::current().id();
+    if !map.contains_key(&thread_id) {
+        map.insert(thread_id, FlashContext::new());
+    }
+    let ctx = map.get_mut(&thread_id).unwrap();
+    ctx.flash_params.insert(dev_id, FlashParamsStruct {
         align: dev.align() as u8,
         erased_val: dev.erased_val(),
     });
-
     let dev: &'static mut dyn Flash = mem::transmute(dev);
-    let mut flash = FLASH.lock().unwrap();
-    flash.insert(dev_id, FlashPtr{ptr: dev as *mut dyn Flash});
+    ctx.flash_map.insert(dev_id, FlashPtr{ptr: dev as *mut dyn Flash});
 }
 
 pub unsafe fn clear_flash(dev_id: u8) {
-    let mut flash = FLASH.lock().unwrap();
-    flash.remove(&dev_id);
+    let mut map = THREAD_CTX.lock().unwrap();
+    let thread_id = thread::current().id();
+    let ctx = map.get_mut(&thread_id).unwrap();
+    ctx.flash_map.remove(&dev_id);
 }
 
 // This isn't meant to call directly, but by a wrapper.
 
+#[no_mangle]
+pub extern fn sim_get_flash_areas() -> *const CAreaDesc {
+    let map = THREAD_CTX.lock().unwrap();
+    let thread_id = thread::current().id();
```

Comment:
Rust std has thread local variables (https://doc.rust-lang.org/std/macro.thread_local.html), I wonder if this would be cleaner to just use those?

Reply:
I'm also curious if making these thread local values will remove the need to derive Send for the wrapper types.

Reply:
I updated it to use `thread_local` there's no need to derive `Send` anymore.